### PR TITLE
WasmPlugin: Make the hard coded parameters in Cache configurable

### PR DESCRIPTION
**Pull request by ingwonsong**
This PR proposes to introduce a set of the environment variables in `istio-agent` for helping the integration tests or e2e tests by making the wasm cache more configurable.

In addition, in the discussion https://github.com/istio/istio/pull/38571#discussion_r861209547, it was also requested.

### CODE REVIEW

File: pkg/wasm/httpfetcher_test.go [hunk 1]
```diff
@@ -84,7 +84,7 @@ func TestWasmHTTPFetch(t *testing.T) {
 				gotNumRequest++
 			}))
 			defer ts.Close()
-			fetcher := NewHTTPFetcher(1 * time.Second)
+			fetcher := NewHTTPFetcher(1*time.Second, defaultOptions().HTTPRequestMaxRetries)
```
Comment (zirain) [inline pkg/wasm/httpfetcher_test.go:87]:
```suggestion
			fetcher := NewHTTPFetcher(defaultOptions().HTTPRequestTimeout, defaultOptions().HTTPRequestMaxRetries)
```

**Comment by zirain** [comment]
> @zirain Could you check out CLA?

I have sign CLA with another email address, can you reset that commit?

**Review by zirain — APPROVED** [review]

File: pilot/cmd/pilot-agent/options/options.go [hunk 2]
```diff
@@ -104,6 +104,18 @@ var (
 	wasmInsecureRegistries = env.RegisterStringVar("WASM_INSECURE_REGISTRIES", "",
 		"allow agent pull wasm plugin from insecure registries or https server, for example: 'localhost:5000,docker-registry:5000'").Get()
 
+	wasmModuleExpiry = env.RegisterDurationVar("WASM_MODULE_EXPIRY", 24*time.Hour,
+		"cache expiration duration for a wasm module.").Get()
+
+	wasmPurgeInterval = env.RegisterDurationVar("WASM_PURGE_INTERVAL", 10*time.Minute,
+		"interval between checking the expiration of wasm modules").Get()
+
+	wasmHTTPRequestTimeout = env.RegisterDurationVar("WASM_HTTP_REQUEST_TIMEOUT", 5*time.Second,
```
Comment (hzxuzhonghu) [inline pilot/cmd/pilot-agent/options/options.go:113]: 5s looks too short, it is downloading a module.   Change it to 15s

File: pkg/wasm/cache.go [hunk 3]
```diff
@@ -127,20 +110,63 @@ type cacheEntry struct {
...
 }
 
+func defaultOptions() CacheOptions {
+	return CacheOptions{
+		PurgeInterval:         10 * time.Minute,
+		ModuleExpiry:          24 * time.Hour,
+		InsecureRegistries:    sets.New(),
+		HTTPRequestTimeout:    5 * time.Second,
+		HTTPRequestMaxRetries: 5,
+	}
```
Comment (hzxuzhonghu) [inline pkg/wasm/cache.go:120]: These should be same as flags

File: pilot/cmd/pilot-agent/options/options.go [hunk 4]
```diff
@@ -104,6 +104,18 @@ var (
 	wasmInsecureRegistries = env.RegisterStringVar("WASM_INSECURE_REGISTRIES", "",
 		"allow agent pull wasm plugin from insecure registries or https server, for example: 'localhost:5000,docker-registry:5000'").Get()
 
+	wasmModuleExpiry = env.RegisterDurationVar("WASM_MODULE_EXPIRY", 24*time.Hour,
+		"cache expiration duration for a wasm module.").Get()
+
+	wasmPurgeInterval = env.RegisterDurationVar("WASM_PURGE_INTERVAL", 10*time.Minute,
```
[the remark this answers is not in the document]
Comment (ingwonsong) [reply pilot/cmd/pilot-agent/options/options.go:110]: I thought that 10min seems to be not so frequent, but longer interval would be also great.
If you have any interval in your mind, could you recommend it for me?

File: pilot/cmd/pilot-agent/options/agent.go [hunk 5]
```diff
@@ -39,6 +39,10 @@ func NewAgentOptions(proxy *model.Proxy, cfg *meshconfig.ProxyConfig) *istioagen
 		EnableDynamicProxyConfig:    enableProxyConfigXdsEnv,
 		EnableDynamicBootstrap:      enableBootstrapXdsEnv,
 		WASMInsecureRegistries:      strings.Split(wasmInsecureRegistries, ","),
+		WASMModuleExpiry:            wasmModuleExpiry,
+		WASMPurgeInterval:           wasmPurgeInterval,
+		WASMHTTPRequestTimeout:      wasmHTTPRequestTimeout,
+		WASMHTTPRequestMaxRetries:   wasmHTTPRequestMaxRetries,
```
[the remark this answers is not in the document]
Comment (ingwonsong) [reply pilot/cmd/pilot-agent/options/agent.go]: Done. they are gathered into wasm.Options

File: pilot/cmd/pilot-agent/options/options.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ingwonsong) [reply pilot/cmd/pilot-agent/options/options.go]: I put 1 hour. If you have other thoughts, please let me know.

File: pilot/cmd/pilot-agent/options/options.go (hunk 2 above)
Comment (ingwonsong) [reply pilot/cmd/pilot-agent/options/options.go]: Done

File: pkg/wasm/cache.go (hunk 3 above)
Comment (ingwonsong) [reply pkg/wasm/cache.go]: Done.

File: pilot/cmd/pilot-agent/options/options.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/cmd/pilot-agent/options/options.go]: SGTM

File: pilot/cmd/pilot-agent/options/agent.go [hunk 6]
```diff
@@ -23,22 +23,30 @@ import (
...
-		WASMInsecureRegistries:      strings.Split(wasmInsecureRegistries, ","),
+		XDSRootCerts:             xdsRootCA,
+		CARootCerts:              caRootCA,
+		XDSHeaders:               map[string]string{},
+		XdsUdsPath:               filepath.Join(cfg.ConfigPath, "XDS"),
+		IsIPv6:                   proxy.IsIPv6(),
+		ProxyType:                proxy.Type,
+		EnableDynamicProxyConfig: enableProxyConfigXdsEnv,
+		EnableDynamicBootstrap:   enableBootstrapXdsEnv,
+		WASMOptions: wasm.Options{
```
Comment (hzxuzhonghu) [inline pilot/cmd/pilot-agent/options/agent.go:43]: You can add a construction function for Options, like NewOption() to replace `defaultOptions`

And then you can not hardcoding the default value in https://github.com/istio/istio/pull/39050/files#diff-be623e52351ff6a4def385a3dce45551df39e74750e7625dc937ae7a1aa1c8bcR107-R117

**Review by hzxuzhonghu — APPROVED** [review]
LGTM, though

**Review by GregHanson — APPROVED** [review]

File: pilot/cmd/pilot-agent/options/agent.go (hunk 6 above)
Comment (ingwonsong) [reply pilot/cmd/pilot-agent/options/agent.go:43]: Just move the values into the constant to avoid hard coding in the two places.

**Pull request merged**